auditbeat/module/file_integrity/kprobes/executor.go

```diff
@@ -0,0 +1,128 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build linux
+
+package kprobes
+
+import (
+	"context"
+	"runtime"
+
+	"golang.org/x/sys/unix"
+)
+
+type executor interface {
+	Run(f func() error) error
+	GetTID() int
+}
+
+// fixedExecutor runs tasks on a fixed OS thread (see runtime.LockOSThread).
+type fixedExecutor struct {
+	ctx      context.Context
+	cancelFn context.CancelFunc
+	// tid is the OS identifier for the thread where it is running.
+	tid  int
+	runC chan func() error
+	retC chan error
+}
+
+// Run submits new tasks to run on the executor and waits for them to finish returning any error.
+func (ex *fixedExecutor) Run(f func() error) error {
+	if ctxErr := ex.ctx.Err(); ctxErr != nil {
+		return ctxErr
+	}
+
+	select {
```

Comment:
why 2 selects?

Reply:
the first one to push the func to be executed in `runC` chan but also get out if ctx is done and the second to get the resulting err from the func we wrote in the first select through `retC` but also get out if ctx is done